prometheus/promhttp/http.go

```diff
@@ -406,3 +435,46 @@ func httpError(rsp http.ResponseWriter, err error) {
 		http.StatusInternalServerError,
 	)
 }
+
+// NegotiateEncodingWriter reads the Accept-Encoding header from a request and
+// selects the right compression based on an allow-list of supported
+// compressions. It returns a writer implementing the compression and an the
+// correct value that the caller can set in the response header.
+func NegotiateEncodingWriter(r *http.Request, rw io.Writer, disableCompression bool, compressions []string) (_ io.Writer, encodingHeaderValue string, closeWriter func() error, _ error) {
```

Comment:
This is good, but I think we can't make it public. I know you wanted to reuse this in other projects (e.g. KSM), but this is not ideal for client_golang. Imagine we will have to maintain random HTTP semantic convention logic until v2 for everybody, including that internal vendored package... which is not that internal, because we have exported function that uses it at the end 🙈 

Sorry... For KSM, I would copy for now (A little copying is better than a little dependency](https://go-proverbs.github.io/#:~:text=A%20little%20copying%20is%20better%20than%20a%20little%20dependency)), OR consider adding this to prometheus/common OR separate repo/project.


Second suggestion - we can kill disableCompression and use empty `compressions` for this case.
3rd suggestion - closeWriter does not need to return error (easier to use)

```suggestion
func negotiateEncodingWriter(r *http.Request, rw io.Writer, compressions []string) (_ io.Writer, encodingHeaderValue string, closeWriter func(), _ error) {
```

Reply:
I understand your points and changed it to private
Do you think prometheus/common would be open to include it?

I saw the internalized autoneg dependency https://github.com/prometheus/common/pull/625 got replaced recently and could probably be replaced by gddo/httputil (but that would mean to move it internally as well).